### PR TITLE
Change endpoint path structure to resourceNames

### PR DESCRIPTION
**Pull request by coderkevin**
This makes a fundamental change in the way data is stored by simplifying all endpoint structure into a flat string indexed structure (thanks @dmsnell!)

Note: This PR actually removes a net of 120 (real) lines of code and 2.1k from the bundle size. 🎉

To Test:
1. `npm install`
2. `npm run test`
3. `npm run lint`
4. `npm link`
5. cd into `examples/hello-world`, `npm install`, `npm link fresh-data`, `npm start` and verify operation.
6. cd into `examples/wp-rest-api`, `npm install`, `npm link fresh-data`, `npm start` and verify operation.

### CODE REVIEW

**Comment by coderkevin** [comment]
Thanks for the review @timmyc! Just noting here that you mentioned a re-clone fixed your lint step and setting this ready to merge.

File: examples/hello-world/src/test-api.js [hunk 1]
```diff
@@ -45,22 +45,22 @@ export default class TestApi extends FreshDataApi {
...
-				const value = get( fullEndpointPath )( params );
-				return value;
-			},
-		},
+	static operations = {
+		read: ( { get } ) => ( resourceNames ) => {
+			return resourceNames.reduce( ( requests, name ) => {
+				if ( 'greetings' === name ) {
+					requests[ name ] = get( [ name ] )();
+				}
```
Comment (dmsnell) [inline examples/hello-world/src/test-api.js]: this is something that isn't where I'd like it with `http-data`. right now the call to `requestHttpData()` must specify the calling action for fetching the data. instead I'd rather have it register a way to get data of a given id and henceforth allow fetching via a call to `get( id )`

that introduces complexity you address here and I'm not sure how best to pull it into `http-data`, but thanks for this!

File: examples/hello-world/src/test-api.js [hunk 2]
```diff
@@ -45,22 +45,22 @@ export default class TestApi extends FreshDataApi {
...
+				return requests;
+			}, {} );
```
Comment (dmsnell) [inline examples/hello-world/src/test-api.js]: what is the purpose of this function? it looks like it's establishing a library of data types that it will then `get` with some passed-in getter, but that getter takes an id as input. if we call this with a `get` parameter anyway, why is this needed? could we not just call `get( 'my-thing' )`?

File: examples/wp-rest-api/src/test-wp-rest-api.js [hunk 3]
```diff
@@ -24,23 +25,23 @@ export function createApi( fetch = window.fetch ) {
...
-					const fullEndpointPath = [ 'posts', ...endpointPath ];
-					const value = get( fullEndpointPath )( params );
-					return value;
-				},
+		static operations = {
+			read: ( { get } ) => ( resourceNames ) => {
+				const postPages = resourceNames.filter( name => startsWith( name, 'posts-page:' ) );
+
+				return postPages.reduce( ( requests, name ) => {
+					const params = JSON.parse( name.substr( name.indexOf( ':' ) + 1 ) );
```
Comment (dmsnell) [inline examples/wp-rest-api/src/test-wp-rest-api.js]: are you encoding a JSON string after the param?

also, for funnies…

```js
const [ type, data ] = name.split( ':' );
const params = JSON.parse( data );
```

File: examples/hello-world/src/test-api.js (hunk 1 above)
Comment (coderkevin) [reply examples/hello-world/src/test-api.js]: I decided to take some inspiration from redux reducers here. Operations can be combined and composed, and a single operation can handle just one resourceName, or many. This becomes more evident in API like in WP REST API where you can pass in `includes=24,25,26,27,28`. This makes it possible to fulfill an operation on multiple resourceNames with one API call, which is one of the goals I was looking for.

In addition, it also allows the operation to save the data in custom ways. An example of this is fetching a page of posts, then storing each post under its own resourceName, and then storing the page as a simple list of IDs. Then the corresponding page selector can grab the page as a list of ids, then map those ids to the actual items for the page. The advantage of this is, the post data can be shared amongst multiple resource requests, such as page requests, search requests, or individual post requests. This also allows the same data to be updated after a mutating operation like an update, and then nothing has to be done to update the data on the page afterwards.

File: examples/hello-world/src/test-api.js (hunk 2 above)
Comment (coderkevin) [reply examples/hello-world/src/test-api.js]: The `get` passed in here is an API method, which is a general-purpose API function. The WP REST API only has two methods: `GET` and `POST`. They are implemented once and passed in to every operation to be used.

This provides an abstraction point where the methods can be easily implemented in different ways on different environments (such as using a proxy in Calypso on wordpress.com, or using OAuth authenticated HTTP requests, or using `apiRequest` from `@wordpress/data`. The methods also could be pointed to another data system for further integration.

Comment (coderkevin) [reply examples/hello-world/src/test-api.js]: >  could we not just call get( 'my-thing' )?

I think my longer explanation [above](https://github.com/coderkevin/fresh-data/pull/56/files/3aeeb9ca840ceea73f3a88b263d37986fe931630#r200397801) explains why one would want to do resource-specific things here in the operations, rather than the methods. Let me know if you still have questions.

File: examples/wp-rest-api/src/test-wp-rest-api.js (hunk 3 above)
Comment (coderkevin) [reply examples/wp-rest-api/src/test-wp-rest-api.js]: Ooh, I like that better, much more tidy! Totally going to use that.
